### ISSUE TITLE
False positive PropertyUsedBeforeDeclaration on good code?

### ISSUE THREAD

**Issue by TWiStErRob** [detekt/detekt#6147]
## Observed Behavior
```kotlin
class C {
	var externalProperty: Boolean
		get() = state
		set(value) {
			state = value
		}

	var derived1: () -> String =
		{
			if (state) "on" else "off"
		}

	fun derived2(): String =
		if (state) "on" else "off"

	//var bad1: String = derived1
	//var bad2: String = derived2()

	private var state: Boolean = false
}
```
all usages of `state` is flagged (including the write usage in `set`), even though there's no way for the execution to produce any different behavior.

## Expected Behavior
No flagging, unless bad1 or bad2 is uncommented.

## Steps to Reproduce
See "Observed"

## Context
In its current state it's unclear to me how this rule is different than
> Detekt - ClassOrdering: property `state` should be declared before method declarations.

other than maybe also working on function level. Should it be restricted to function level? Because cross-function flow analysis might be a bit tricky.

## Your Environment

* Version of detekt used: 1.23.0
* Version of Gradle used (if applicable): 8.1.1
* Gradle scan link (add `--scan` option when running the gradle task): N/A
* Operating System and version: all
* Link to your project (if it's a public repository): N/A

**Comment by cortinico** [comment]
> No flagging, unless bad1 or bad2 is uncommented.

I believe this will be hard to achieve, imho.

**Comment by TWiStErRob** [comment]
> > No flagging, unless bad1 or bad2 is uncommented.

> I believe this will be hard to achieve, imho.

I agree, does that mean the current PSI / Detekt pairing might be too weak to support this rule? Because it would generate too many false positives, and in the end it would just impose the same rule as `ClassOrdering` does. What are your thoughts on this?

**Comment by cortinico** [comment]
> Because it would generate too many false positives, and in the end it would just impose the same rule as ClassOrdering does. What are your thoughts on this?

I feel that this rule is sort of a specialization of `ClassOrdering` which works also on functions and other blocks where properties are allowed.

**Comment by TWiStErRob** [comment]
I'm starting to see it now going through more and more violations: on the class level it's ClassOrdering but ignoring functions and **enforcing order between properties only**.

Note that the order enforced is "lower level to higher level", which is the opposite to what I've seen big companies do for functions ordering: public entry point fun, private member funs, private static utilities.

**Comment by BraisGabin** [comment]
I would like to know the thoughts of @t-kameyama on this issue. No to implement/fix it but because I'm not sure if this is even doable on the current state of detekt. Can we determine the execution flows to distinguish between one case and the other? And, if it is not possible, what can we do to support it?

**Comment by TWiStErRob** [comment]
I think a visitor for the whole class could be used with type resolution to create a mapping from each property/function to what property/function they use (maybe call). After that point it should be a simple reverse lookup / transitive closure on the call graph.

**Comment by t-kameyama** [comment]
I think it is difficult to analyze the execution flow. A simple rule that detects if a property is used before it is defined would be fine. I also think this rule would be useful because ClassOrdering rule does not detect the following problems.

https://github.com/detekt/detekt/issues/749
```
class Foo {
    val number
        get() = if (isValid) 1 else 0
    val list = listOf(number)
    val isValid = true
}
println(Foo().list)
```
